### PR TITLE
Webpack into native module

### PR DESCRIPTION
**Pull request by samholmes**
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

 none

### Description

 none

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207103682121388

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+def safeExtGet(prop, fallback) {
+  rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
+}
+
+def DEFAULT_COMPILE_SDK_VERSION = 28
+def DEFAULT_BUILD_TOOLS_VERSION = '28.0.2'
+def DEFAULT_MIN_SDK_VERSION = 19
+def DEFAULT_TARGET_SDK_VERSION = 27
+
+android {
```
Comment (swansontec) [inline android/build.gradle:23]: Please add the package name in here: ` namespace "app.edge.reactnative.currencyplugins"`, and then remove it from the Android Manifest.

Putting the namespace in the manifest works, but is deprecated.

File: android/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+          package="app.edge.reactnative.currencyplugins">
```
Comment (swansontec) [inline android/src/main/AndroidManifest.xml:2]: We can remove the `package=` prop once we add it to the build.grade.

File: webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+const webpack = require('webpack')
+
+// Run `yarn start.plugins` to enable debug mode.
+// This mode will serve the plugin bundle via a local dev-server.
+const debug = process.env.WEBPACK_SERVE
+
+// Try exposing our socket to adb (errors are fine):
+if (process.env.WEBPACK_SERVE) {
+  console.log('adb reverse tcp:8101 tcp:8101')
+  exec('adb reverse tcp:8101 tcp:8101', () => {})
```
Comment (swansontec) [inline webpack.config.js:14]: This port, 8101, the the one the GUI currently uses. That is fine!

However, edge-core-js uses 8080, edge-currency-accountbased used 8082, and edge-exchange plugins uses 8083. So it seems like 8081 would "fit" better than 8101.

File: webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
+const { exec } = require('child_process')
+const path = require('path')
+const { EsbuildPlugin } = require('esbuild-loader')
+
+const webpack = require('webpack')
+
+// Run `yarn start.plugins` to enable debug mode.
+// This mode will serve the plugin bundle via a local dev-server.
```
Comment (swansontec) [inline webpack.config.js:8]: This yarn instruction is incorrect.

File: webpack.config.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      https: require.resolve('https-browserify'),
+      os: require.resolve('os-browserify/browser'),
+      path: require.resolve('path-browserify'),
+      stream: require.resolve('stream-browserify'),
+      string_decoder: require.resolve('string_decoder'),
+      url: require.resolve('url'),
+      vm: require.resolve('vm-browserify')
+    },
+    mainFields: ['browser', 'module', 'main']
+  }
```
Comment (swansontec) [inline webpack.config.js:83]: Now that this config file only needs to worry about edge-currency-plugins and not anybody else, I'm guessing we can really minimize this section. Does edge-currency-plugins really depend on all this Node.js stuff? Probably not all of it.

However, I understand the motivation to keep things stable for now. Maybe we can create a follow-on task and handle this cleanup in a future release.

File: README.md [hunk 6]
```diff
@@ -1,3 +1,62 @@
...
+
+// Once you are done adding plugins, call this:
+lockEdgeCorePlugins();
+```
+
+You can also add plugins individually if you want to be more picky:
+
+```js
+addEdgeCorePlugins({
+  ethereum: plugins.ethereum,
```
Comment (swansontec) [inline README.md:31]: Probably a bad example for this repo.

File: README.md [hunk 7]
```diff
@@ -1,3 +1,62 @@
...
+### Browser
+
+The bundle located in `dist/edge-currency-plugins.js` will automatically register itself with edge-core-js. Just serve the entire `dist` directory along with your app, and then load the script:
+
+```html
+<script src='https://example.com/app/dist/edge-currency-plugins.js'>
+```
+
+If you want to debug this project, run `yarn start` to start a Webpack server,
+and then adjust your script URL to http://localhost:8101/edge-currency-plugins.js.
```
Comment (swansontec) [inline README.md:44]: If we do adjust the port number, change it here too.

File: src/react-native.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { EdgeOtherMethods } from 'edge-core-js/types'
+import { NativeModules } from 'react-native'
+// import { bridgifyObject, emit, onMethod } from 'yaob'
+
+const { EdgeCurrencyPluginsModule } = NativeModules
+const { sourceUri } = EdgeCurrencyPluginsModule.getConstants()
+
+export const pluginUri = sourceUri
+export const debugUri = 'http://localhost:8101/edge-currency-plugins.js'
```
Comment (swansontec) [inline src/react-native.ts:8]: If we do change the port number, adjust it here too.

**Review by swansontec — CHANGES_REQUESTED** [review]
Just a few small things need to get fixed.

File: android/src/main/AndroidManifest.xml (hunk 2 above)
Comment (samholmes) [reply android/src/main/AndroidManifest.xml:2]: Hmm, this is not how it is configured in accountbased.

File: android/build.gradle (hunk 1 above)
Comment (samholmes) [reply android/build.gradle:23]: Hmm, this is not how it is configured in accountbased.

Comment (swansontec) [reply android/build.gradle:23]: See edge-core-js for an example. I guess we need to update accountbased too.

**Review by swansontec — APPROVED** [review]

**Pull request merged**